cEP-0018.md

```diff
@@ -0,0 +1,276 @@
+        """
+        ret_class = make_walker(lang)
+        if not ret_class:
+            raise Exception('Un-supported Language')
+        else:
+            return ret_class(file, filename)
+
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """
+        Modify the internal tree variable and return the next node
+        """
+
+    def get_prev_node(self):
+        """
+        Modify the internal tree variable and return prev node
+        """
+
+    def reset(self):
+        """
+        Reset the internal tree variable to tree root
+        """
+
+    def is_terminal_node(node):
+        """
+        Determine if the node is terminal
+        """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+    ...
+```
+
+### Prototype of Py3ASTWalker
+
+These kinds of Walkers will supply a high level walker api
+
+```python
+class Py3ASTWalker(ASTWalker):
+    def next_function():
+        # modify the tree node variable such that it only stops at functions
+        # which can be detected by the Python grammar rules
+    ...
+```
+
+### Prototype of  `ASTBear` class implementation
+
+```python
+from coantlib import ASTWalker
+from coalib.bears.LocalBear import Bear
+
+class ASTBear(LocalBear):
+    walker = None
+
+    def initialise(file, filename):
+        if LANGUAGES:
+            self.walker = ASTWalker.get_walker(file,
+                                               filename,
+                                               self.LANGUAGES[0])
+        else:
+            self.walker = ASTWalker.get_walker(file, filename, 'auto')
+
+    def run(self,
+            filename,
+            file,
+            tree,
+            *args,
+            dependency_results = None,
+            **kwargs
+            ):
+        raise NotImplementedError('Needs to be done by bear')
+```
+
+### Prototype of  `Py3ASTBear` class implementation
+
+These kinds of base bears will provide an even more abstracted high level api
+Following is a demonstration in case language grammars have common things in them
+(For e.g Python2 and Python3)
+
+```python
+from coantlib import ASTBear
+
+class BasePyASTBear(ASTBear):
+    LANGUAGES = { 'Python' }
+    def get_imports():
+        """
+        Return list of strings containing import statements from the file
+        """
+
+    def get_methods():
+        """
+        Return list of strings containing all methods from the file
+        """
+    ...
+
+class Py2ASTBear(BasePyASTBear):
+    LANGUAGES = { 'Python 2' }
+    def get_print_statements():
+        """
+        Return list of print statements, since print is a keyword in py2
+        """
+    ...
+
+class Py3ASTBear(BasePyASTBear):
+    LANGUAGES = { 'Python3' }
+    def get_integer_div_stmts():
+        """
+        Returns a list of integer divide statements, since / and // are
+        different in py3
+        """
+    ...
+```
```

Comment:
I think you have got it wrong. Bears have to provide some results. Run some analysis. High level APIs like `get_methods` and `get_imports` should be part of the `Py3ASTWalker`. If then you want this walker returned then say `SampleBear` is a python3 bear then it will inherit `ASTBear`. Then in the run method of `SampleBear` one calls something like `get_ast_walker` then that will be a  method in `ASTBear` which will look at the `LANGUAGES` parameter of `SampleBear` and automatically return the right walker which should be the `Py3ASTWalker`. We will have a seperate file which maps a `LanguageDefinition` which already exists in `bearlib` in coala core repository to a relevant walker. Thus `bearlib` would help us resolved the walker through the language identification methods it already has.

This way the bear writer doesn't have to worry about inheriting the right bear.